SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp

```diff
@@ -0,0 +1,281 @@
+
+    return log_msg;
+  }
+
+private:
+  template <init_val Val> std::string init_val_to_string() const {
+    if constexpr (Val == init_val::min) {
+      return "min";
+    } else if constexpr (Val == init_val::max) {
+      return "max";
+    } else if constexpr (Val == init_val::zero) {
+      return "zero";
+    } else if constexpr (Val == init_val::positive) {
+      return "positive";
+    } else if constexpr (Val == init_val::negative) {
+      return "negative";
+    } else if constexpr (Val == init_val::min_half) {
+      return "min_half";
+    } else if constexpr (Val == init_val::max_half) {
+      return "max_half";
+    } else if constexpr (Val == init_val::neg_inf) {
+      return "neg_inf";
+    } else if constexpr (Val == init_val::nan) {
+      return "nan";
+    } else if constexpr (Val == init_val::denorm) {
+      return "denorm";
+    } else if constexpr (Val == init_val::inexact) {
+      return "inexact";
+    } else if constexpr (Val == init_val::ulp) {
+      return "ulp";
+    } else if constexpr (Val == init_val::ulp_half) {
+      return "ulp_half";
+    } else {
+      static_assert(Val != Val, "Unexpected base value value");
+    }
+  }
+};
+
+template <typename DataT, int NumElems, typename TestCaseT, typename BaseVal,
+          typename Step>
+class run_test {
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+    static_assert(std::is_same_v<typename BaseVal::value_type, init_val>,
+                  "BaseVal template parameter should be init_val type.");
+    static_assert(std::is_same_v<typename Step::value_type, init_val>,
+                  "Step template parameter should be init_val type.");
+
+    shared_vector<DataT> result(NumElems, shared_allocator<DataT>(queue));
+
+    const auto base_value = get_value<DataT, BaseVal::value>();
+    const auto step_value = get_value<DataT, Step::value>(base_value);
+
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *const out = result.data();
+
+      cgh.single_task<kernel_for_fill<DataT, NumElems, TestCaseT,
+                                      BaseVal::value, Step::value>>(
+          [=]() SYCL_ESIMD_KERNEL {
+            TestCaseT::template call_simd_ctor<DataT, NumElems>(
+                base_value, step_value, out);
+          });
+    });
+    queue.wait_and_throw();
+
+    bool passed = true;
+
+    DataT expected_value = base_value;
+    for (size_t i = 0; i < result.size(); ++i) {
+      // std::isnan() couldn't be called for integral types because it call is
+      // ambiguous GitHub issue for that case:
+      // https://github.com/microsoft/STL/issues/519
+      if (!are_bitwise_equal(result[i], expected_value)) {
+        if constexpr (type_traits::is_sycl_floating_point_v<DataT>) {
+          if (std::isnan(result[i]) &&
+              (std::isnan(step_value) || std::isnan(base_value))) {
+            continue;
+          }
+        }
+        passed = false;
```

Comment:
```suggestion
    if constexpr ((Base == init_val::nan) || (Step == init_val::nan)) {
      if (!are_bitwise_equal(result[0], base_value))) {
        ...
      }
      for (size_t i = 1; i < result.size(); ++i) {
        if (!isnan(result[i])) {
          ...
        }
      }
    } else {
      DataT expected_value = base_value;
      for (size_t i = 0; i < result.size(); ++i) {
        if (!are_bitwise_equal(result[i], expected_value)) {
          passed = false;
```
with
```
  constexpr Base = BaseValT::value;
  constexpr Step = StepT::value;
```

Otherwise we would miss the difference in opcode for `result[0]`

Reply:
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/209a007ff239c59795d58172a05bf77ba4fd5b80#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4L224-L244)